catalog/app/utils/Dialogs.tsx

```diff
@@ -0,0 +1,67 @@
+import * as React from 'react'
+import * as M from '@material-ui/core'
+
+import defer from 'utils/defer'
+import type { Resolver } from 'utils/defer'
+
+type DialogState = 'open' | 'closing' | 'closed'
+
+type ExtraDialogProps = Omit<M.DialogProps, 'open' | 'onClose' | 'onExited'>
+
+export type Close<R> = [R] extends [never] ? () => void : Resolver<R>['resolve']
```

Comment:
I don't understand this syntax :(

Reply:
this might be helpful: https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

basically i wanted to allow calling the `close()` function without any arguments by default, so i used a conditional type here, but it wasn't working out of the box for some cases until i used the workaround described in the docs

Reply:
Thanks!